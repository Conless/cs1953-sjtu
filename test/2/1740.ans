7
1
5
5
10
1
10
10
1
8
