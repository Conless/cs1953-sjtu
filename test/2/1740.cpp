#include <bits/stdc++.h>

using namespace std;

int n, m, q;
int vis[100005];
vector<pair<int, int> > edge[100005];
queue<int> qq, qqq;

bool check(int x, int y, int z, int dis) {
	while (!qq.empty())
		qq.pop();
	while (!qqq.empty())
		qqq.pop();
	qq.push(x);
	qq.push(y);
	qqq.push(x);
	qqq.push(y);
	vis[x] = vis[y] = 1;
	bool fl = 0;
	while (!qq.empty()) {
		if (qqq.size() >= z) {
			fl = 1;
			break;
		}
		int u = qq.front();
		qq.pop();
		int llen = edge[u].size();
		for (int i = 0; i < llen; i++) {
			if (edge[u][i].first > dis)
				break;
			int v = edge[u][i].second;
			if (!vis[v]) {
				qq.push(v);
				vis[v] = 1;
				qqq.push(v);
			}
		}
	}
	while (!qqq.empty()) {
		int u = qqq.front();
		vis[u] = 0;
		qqq.pop();
	}
	return fl;
}

int main() {
    freopen("1740.in", "r", stdin);
    freopen("1740.ans", "w", stdout);
	scanf("%d%d", &n, &m);
	for (int i = 1; i <= m; i++) {
		int x, y;
		scanf("%d%d", &x, &y);
		edge[x].push_back(make_pair(i, y));
		edge[y].push_back(make_pair(i, x));
	}
	scanf("%d", &q);
	while (q--) {
		int x, y, z;
		scanf("%d%d%d", &x, &y, &z);
		int l = 1, r = m, ans = m + 1;
		while (l <= r) {
			int mid = (l + r) >> 1;
			if (check(x, y, z, mid)) {
				ans = mid;
				r = mid - 1;
			} else l = mid + 1;
		}
		printf("%d\n", ans);
	}
	return 0;
}