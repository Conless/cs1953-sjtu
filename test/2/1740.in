10 20
2 1
2 2
7 4
2 6
5 2
4 8
9 8
7 2
5 4
10 9
9 2
6 2
2 4
6 2
9 2
5 10
3 7
4 9
2 5
6 8
10
7 1 8
7 10 2
10 5 3
9 5 3
10 10 7
4 4 2
5 4 9
1 5 9
9 10 2
2 5 5
