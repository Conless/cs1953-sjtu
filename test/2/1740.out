7
1
5
5
10
3
10
10
1
8
